libobs/obs-encoder.c

```diff
@@ -1874,3 +1879,93 @@ uint64_t obs_encoder_get_pause_offset(const obs_encoder_t *encoder)
 {
 	return encoder ? encoder->pause.ts_offset : 0;
 }
+
+bool obs_encoder_has_roi(const obs_encoder_t *encoder)
+{
+	return encoder->roi.num > 0;
+}
+
+bool obs_encoder_add_roi(obs_encoder_t *encoder,
+			 const struct region_of_interest *roi)
+{
+	if (!roi || !encoder->media)
+		return false;
+	if (!(encoder->info.caps & OBS_ENCODER_CAP_ROI))
+		return false;
+	/* Area smaller than the smallest possible block (16x16) */
+	if (roi->bottom - roi->top < 16 || roi->right - roi->left < 16)
+		return false;
+	/* Other invalid ROIs */
+	if (roi->priority < -1.0f || roi->priority > 1.0f)
+		return false;
+
+	pthread_mutex_lock(&encoder->roi_mutex);
+
+	if (encoder->scaled_height && encoder->scaled_width) {
+		const uint32_t width = video_output_get_width(encoder->media);
+		const uint32_t height = video_output_get_height(encoder->media);
+
+		if (width && height) {
+			/* Scale ROI to output size */
+			const float scale_x =
+				(float)encoder->scaled_width / (float)width;
+			const float scale_y =
+				(float)encoder->scaled_height / (float)height;
+
+			const struct region_of_interest scaled_roi = {
+				.top = (uint32_t)((float)roi->top * scale_y),
+				.bottom = (uint32_t)((float)roi->bottom *
+						     scale_y),
+				.left = (uint32_t)((float)roi->left * scale_x),
+				.right =
+					(uint32_t)((float)roi->right * scale_x),
+				.priority = roi->priority,
+			};
+
+			da_push_back(encoder->roi, &scaled_roi);
+			encoder->roi_increment++;
+		}
+	} else {
+		da_push_back(encoder->roi, roi);
+		encoder->roi_increment++;
+	}
+
+	pthread_mutex_unlock(&encoder->roi_mutex);
+
+	return true;
+}
+
+void obs_encoder_remove_roi(obs_encoder_t *encoder,
+			    const struct region_of_interest *roi)
+{
+	pthread_mutex_lock(&encoder->roi_mutex);
+	da_erase_item(encoder->roi, roi);
```

Comment:
We need to handle the case that the scale is enabled like `obs_encoder_add_roi` does.

Reply:
It looks the force-pushed commit resolves this.

Reply:
It's not a force push yet because I'm not entirely sure if that's the appropriate way to solve this either. I'm trying to figure out if it should be expected for ROIs passed to the callback to work for the remove call, or if the caller should be responsible of keeping track of what ROIs they added. Perhaps there should be a separate means of enumerating all added ROIs and receiving properly scaled ones for use within the encoder.